http-api/src/v1/projects.rs

```diff
@@ -160,37 +159,56 @@ async fn project_root_handler(
 /// `GET /projects/:project/commits/:sha`
 async fn commit_handler(
     Extension(ctx): Extension<Context>,
-    Path((project, sha)): Path<(Urn, One)>,
+    Path((project, sha)): Path<(Urn, Oid)>,
 ) -> impl IntoResponse {
-    let reference = Reference::head(Namespace::from(project.clone()), None, sha.to_owned());
-    let commit = browse(reference, ctx.paths.clone(), |browser| {
-        let oid = browser.oid(&sha)?;
-        radicle_source::commit(browser, oid)
-    })
-    .await?;
-
     let projects = ctx.projects.read().await;
+
+    let repo = git::Repository::new(ctx.paths.git_dir()).map_err(Error::from)?;
+    let browser = git::Browser::new_with_namespace(
+        &repo,
+        &git::Namespace::try_from(project.encode_id().as_str())
+            .map_err(|_| Error::MissingNamespace)?,
+        *sha,
+    )
+    .map_err(Error::from)?;
+
+    let history = browser.get();
+    let commit = history.first();
+
+    let diff = if let Some(parent) = commit.parents.first() {
+        browser.diff(*parent, *sha)?
+    } else {
+        browser.initial_diff(*sha)?
+    };
+
     let fingerprints = projects.get(&project);
-    let fp = ctx.commit_ssh_fingerprint(&commit.header.sha1.to_string())?;
+    let fp = ctx.commit_ssh_fingerprint(&commit.id.to_string())?;
     let committer = if let (Some(fps), Some(fp)) = (fingerprints, fp) {
         fps.get(&fp).cloned().map(|peer| Committer { peer })
     } else {
         None
     };
 
-    let response = Commit {
-        header: commit.header,
-        diff: commit.diff,
-        stats: commit.stats,
-        branches: commit.branches,
+    let branches = browser
+        .revision_branches(*sha)?
+        .iter()
```

Comment:
You can `into_iter` to take ownership of the data and you don't need the `cloned` then.

Reply:
Done